### DISCUSSION TITLE
How to parse html for replacement tokens with a transform

### DISCUSSION THREAD

**Discussion by madsem** [11ty/eleventy#2201] [Q&A] [answered]
Hey all,

I'm trying to parse the text content of the generated html in `dist/`, for tokens that need to be replaced.
I'm using the Netlify CMS, and want to allow writers to use tokens in their articles that will later be replaced.

For example if a writer would write "We have a big selection of handmade {keyword:chocolate} pralines."
I would like to replace the `{keyword:chocolate}` token with an alpinejs html snippet, something like this: `<span x-text="query.keyword"> chocolate </span>`.

What this would do is, if there is a url parameter by the name `&keyword=truffel`,
the sentence would read: "We have a big selection of handmade truffel pralines.", or fall back to the default "chocolate".

Not 100% sure how I can achieve this, if I just scan the entire string that is passed to my transform, I will have issues properly replacing multiple of these occurrences.

So I figured I need to parse the html, somehow loop through every node and check if it contains a token like the above, then replace it as above.

Or maybe I am completely overthinking how this can be done with 11ty?

Hoping one of you can point me in the right direction :)

So far I have:

### 11ty.js
```js
// Parse dynamic replacement tokens for inserting url parameter values into page content.
config.addTransform("replaceQueryTokens", require("./11ty/transforms/replace-query-tokens"));
```

### ./11ty/transforms/replace-query-tokens.js
```js
const { parse, HTMLElement } = require("node-html-parser");

module.exports = (content, outputPath = '.html') => {

    if ( !String(outputPath).endsWith('.html') ) return content;

    return _replaceContentTokens(content)

}

/**
     * Replace all dynamic url parameter insertion tokens in content, with alpinejs enabled html.
     * Example token: {state:Your Area}
     * 
     * Will be replaced with the following alpinejs code: <span x-text="query.state">Your Area</span>
     * 
     * site/_includes/layouts/base.njk initializes the x-data object "query" from the url query string.
     *
     * @param content
     * @returns {*}
     * @private
     */
 function _replaceContentTokens(content) {

    console.log(parse(content))
}
```

**Comment by pdehaan** [comment, marked as answer]
@madsem Possibly have two solutions&hellip;

Here's my transform (which uses a small regex w/ global replace in case there are multiple keywords per output file):

```js
module.exports = (content, outputPath = ".html") => {
  if (!String(outputPath).endsWith(".html")) return content;
  return _replaceContentTokens(content);
};

function _replaceContentTokens(content) {
  return content.replace(/{keyword:(.*?)}/g, `<span x-text="query.keyword">$1</span>`);
}
```

And my test page:
```njk
---
title: test page
---

<h1>{{ title }}</h1>

<h2>Transform test</h2>
<p>We have a big selection of handmade {keyword:chocolate} pralines.</p>
<p>We have a small selection of handmade {keyword:banana} dates.</p>
<p>We have an average selection of handmade {keyword:salted} turtles.</p>

---

<h2>Shortcode test</h2>
<p>We have a reduced selection of handmade {% keyword "haunted" %} oats.</p>
```

And my output:
```html
<h1>test page</h1>

<h2>Transform test</h2>
<p>We have a big selection of handmade <span x-text="query.keyword">chocolate</span> pralines.</p>
<p>We have a small selection of handmade <span x-text="query.keyword">banana</span> dates.</p>
<p>We have an average selection of handmade <span x-text="query.keyword">salted</span> turtles.</p>

---

<h2>Shortcode test</h2>
<p>We have a reduced selection of handmade <span x-text="query.keyword">haunted</span> oats.</p>
```

The second solution (which might not work for your use cases was that second bit above in the test page, a custom shortcode.
Here's my .eleventy.js file:

```js
module.exports = (eleventyConfig) => {
  eleventyConfig.addTransform("token-transformer", require("./11ty/keyword-transform"));

  // This bit right here. Modify in source instead of using a secondary transformer.
  // Usage: `{% keyword "chocolate" %}`
  // Output: `<span x-text="query.keyword">chocolate</span>`
  eleventyConfig.addShortcode("keyword", function (value = "") {
    return `<span x-text="query.keyword">${value}</span>`;
  });

  return {
    dir: {
      input: "src",
      output: "www",
    },
  };
};
```

**Comment by madsem** [comment]
@pdehaan thanks that already looks great!

I should have mentioned that the tokens are completely dynamic, so a writer can use the parameter name of any query string parameter, or also multiple different ones in one sentence.

Like We have a big selection of handmade {keyword:chocolate} pralines at our store location in {state:Your Area}.

Also, the writers use Netlify CMS, so the tokens appear in the frontmatter not as template tags.
Would it be better in this case to make it a 11ty `filter` and just make sure everywhere this content will appear to add `{{ content | tokenTransform }}` or so?



example markdown file created by netlify cms:
```markdown
---
title: test page
hero_text: "We have a big selection of handmade {keyword:chocolate} pralines at our store location in {state:Your Area}."
---
```

**Reply by pdehaan** [reply]
Yeah, I think that probably eliminates the possibility of using shortcodes, since those would need to be predefined. For example, I was experimenting w/ the shortcode route, but don't think it will fit your use case.

```js
// .eleventy.js
  eleventyConfig.addShortcode("keyword", function (value = "") {
    return `<span x-text="query.keyword">${value}</span>`;
  });
  eleventyConfig.addShortcode("state", function (value = "") {
    return `<span x-text="query.state">${value}</span>`;
  });

  // OR define them in a loop.

  for (const token of ["keyword", "state"]) {
    eleventyConfig.addShortcode(token, function (value = "") {
      return `<span x-text="query.${token}">${value}</span>`;
    });
  }
```

As for the transform route, this seems to work:

```js
// ./11ty/transforms/keyword-transform.js
module.exports = (content, outputPath = ".html") => {
  if (!String(outputPath).endsWith(".html")) return content;
  return _replaceContentTokens(content);
};

function _replaceContentTokens(content) {
  return content
    // Replace keyword+state tokens...
    .replace(/{keyword:(.*?)}/g, `<span x-text="query.keyword">$1</span>`)
    .replace(/{state:(.*?)}/g, `<span x-text="query.state">$1</span>`)
    // Replace any token, but may be riskier.
    .replace(/{(\w+?):(.*?)}/g, `<span x-text="query.$1">$2</span>`);
}
```

And that generic token replacer at the end seemed to convert my `<p>We also have {generic:codes}</p>` token into `<p>We also have <span x-text="query.generic">codes</span></p>`. :shrug:

---

I like your custom filter idea though. Might be safer since you can control which blocks of code get tokenized (vs parsing the entire final source and risking it causing issues w/ inline JavaScript or whatever).

My custom filter looks like this:

```js
  eleventyConfig.addFilter("tokenTransform", function (content = "") {
    return content
      .replace(/{(\w+?):(.*?)}/g, `<span x-text="query.$1">$2</span>`);
  });
```

And my [Nunjucks] layout file looks like:

```njk
<main>{{ content | tokenTransform | safe }}</main>
```

**Reply by madsem** [reply]
This works perfectly, thank you @pdehaan 
And MUCH easier solution than what I had concocted 👍
